### PR TITLE
Merge PnP feature into main

### PR DESCRIPTION
**Pull request by jspaith**
Enable PnP client into the main branch.  New APIs in `iothub_client_properties.h` and add new API's to more natively support PnP constructs in device and module clients.  Needed a few core changes for subscribe to commands, but vast majority of work is fortunately isolated in the `iothub_client_properties.c` and in rewriting the samples to use the API instead of the sample helper.

Note to reviewers - each PR has already been reviewed in the `feature/iot_pnp_stage` branch, so in theory this should not require a full re-review.  I've spent last few days also reviewing it myself and doing fixups here and there (e.g. #2331).

### CODE REVIEW

File: iothub_client/samples/pnp/common/pnp_dps_ll.c [hunk 1]
```diff
@@ -2,58 +2,68 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
```
Comment (danewalton) [inline iothub_client/samples/pnp/common/pnp_dps_ll.c:2]: Should this have a deprecated warning as well? Same for other files in this directory.

File: iothub_client/tests/iothubmoduleclient_ll_ut/iothubmoduleclient_ll_ut.c [hunk 2]
```diff
@@ -400,4 +411,70 @@ TEST_FUNCTION(IoTHubModuleClient_LL_SendMessageDisposition_Test)
...
+
+    //act
+    IOTHUB_CLIENT_RESULT result = IoTHubModuleClient_LL_GetPropertiesAndSubscribeToUpdatesAsync(TEST_IOTHUB_MODULE_CLIENT_LL_HANDLE, TEST_PROPERTIES_RECEIVED_CALLBACK, NULL);
+
+    //assert
+    ASSERT_IS_TRUE(result == IOTHUB_CLIENT_OK);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+}
+
+END_TEST_SUITE(iothubmoduleclient_ll_ut)
```
Comment (danewalton) [inline iothub_client/tests/iothubmoduleclient_ll_ut/iothubmoduleclient_ll_ut.c:480]: nit sorry: newline

**Review by danewalton — APPROVED** [review]

File: iothub_client/samples/pnp/common/pnp_device_client_ll.h [hunk 3]
```diff
@@ -1,6 +1,8 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+// THIS FILE IS DEPRECATED.  SEE ./readme.md FOR MIGRATION INSTRUCTIONS.
```
[the remark this answers is not in the document]
Comment (jspaith) [reply iothub_client/samples/pnp/common/pnp_device_client_ll.h:4]: Yes - I'm leaving the old sample helpers around, but ideally folks are on the cleaner and thoroughly tested API's.

File: iothub_client/samples/pnp/common/pnp_dps_ll.c (hunk 1 above)
Comment (jspaith) [reply iothub_client/samples/pnp/common/pnp_dps_ll.c:2]: Good question but on this one, no.  The core of the DPS stuff here is mostly the C version of `async await` for the DPS registration to complete which isn't PnP related.  Both the simple thermostat and temp controller use this common code still even after rework.

File: iothub_client/tests/iothubmoduleclient_ll_ut/iothubmoduleclient_ll_ut.c (hunk 2 above)
Comment (jspaith) [reply iothub_client/tests/iothubmoduleclient_ll_ut/iothubmoduleclient_ll_ut.c:480]: No need to apologize on nits!  Do I need a newline though?  I looked at the original file (including telling diff to NOT ignore blanks) and it looks the same as before my change.  Is the red minus sign on this an artifact of Git web diffing or is there something I'm missing?

Comment (danewalton) [reply iothub_client/tests/iothubmoduleclient_ll_ut/iothubmoduleclient_ll_ut.c:480]: maybe. As long as the file has it its all good. I trust you haha

**Pull request merged**